File: python/src/totalreclaw/protobuf.py
"""Minimal protobuf wire format encoder for TotalReclaw fact payloads.

Hand-rolled encoder matching server/proto/totalreclaw.proto field layout.
No external protobuf library dependency -- uses only stdlib.

Field numbers match TotalReclawFact:
  1: id (string), 2: timestamp (string), 3: owner (string),
  4: encrypted_blob (bytes), 5: blind_indices (repeated string),
  6: decay_score (double), 7: is_active (bool), 8: version (int32),
  9: (removed in v3 -- now encrypted inside field 4),
  10: content_fp (string),
  11: (removed in v3 -- now encrypted inside field 4),
  12: sequence_id (int64 -- assigned by subgraph, not client),
  13: encrypted_embedding (string)
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional


# ---------------------------------------------------------------------------
# Wire format primitives
# ---------------------------------------------------------------------------


def encode_varint(value: int) -> bytes:
    """Encode a non-negative integer as a protobuf base-128 varint."""
    result = bytearray()
    v = value & 0xFFFFFFFF  # treat as unsigned 32-bit
    while v > 0x7F:
        result.append((v & 0x7F) | 0x80)
        v >>= 7
    result.append(v & 0x7F)
    return bytes(result)


def _write_string(parts: list[bytes], field_number: int, value: str) -> None:
    """Write a string field (wire type 2 = length-delimited)."""
    if not value:
        return
    data = value.encode("utf-8")
    key = (field_number << 3) | 2
    parts.append(encode_varint(key))
    parts.append(encode_varint(len(data)))
    parts.append(data)


def _write_bytes(parts: list[bytes], field_number: int, value: bytes) -> None:
    """Write a bytes field (wire type 2 = length-delimited)."""
    key = (field_number << 3) | 2
    parts.append(encode_varint(key))
    parts.append(encode_varint(len(value)))
    parts.append(value)


def _write_double(parts: list[bytes], field_number: int, value: float) -> None:
    """Write a double field (wire type 1 = 64-bit fixed)."""
    key = (field_number << 3) | 1
    parts.append(encode_varint(key))
    parts.append(struct.pack("<d", value))


def _write_varint_field(parts: list[bytes], field_number: int, value: int) -> None:
    """Write a varint field (wire type 0)."""
    key = (field_number << 3) | 0
    parts.append(encode_varint(key))
    parts.append(encode_varint(value))


# ---------------------------------------------------------------------------
# Fact payload dataclass
# ---------------------------------------------------------------------------


@dataclass
class FactPayload:
    """Client-side fact payload ready for protobuf encoding.

    Mirrors the TypeScript ``FactPayload`` interface in
    ``mcp/src/subgraph/store.ts``.
    """

    id: str
    timestamp: str
    owner: str
    encrypted_blob: str  # Hex-encoded XChaCha20-Poly1305 ciphertext
    blind_indices: list[str] = field(default_factory=list)
    decay_score: float = 1.0
    source: str = ""
    content_fp: str = ""
    agent_id: str = ""
    encrypted_embedding: Optional[str] = None


# ---------------------------------------------------------------------------
# Encoding functions
# ---------------------------------------------------------------------------


def encode_fact_protobuf(fact: FactPayload) -> bytes:
    """Encode a :class:`FactPayload` as minimal protobuf wire format.

    The output is byte-compatible with the TypeScript ``encodeFactProtobuf``
    in ``mcp/src/subgraph/store.ts`` and can be decoded by any standard
    protobuf parser using ``server/proto/totalreclaw.proto``.
    """
    parts: list[bytes] = []

    _write_string(parts, 1, fact.id)
    _write_string(parts, 2, fact.timestamp)
    _write_string(parts, 3, fact.owner)
    _write_bytes(parts, 4, bytes.fromhex(fact.encrypted_blob))

    for index in fact.blind_indices:
        _write_string(parts, 5, index)

    _write_double(parts, 6, fact.decay_score)
    _write_varint_field(parts, 7, 1)  # is_active = true
    _write_varint_field(parts, 8, 3)  # version = 3 (source/agent_id now encrypted inside field 4)
    # Fields 9 (source) and 11 (agent_id) removed in v3 -- now encrypted inside field 4
    _write_string(parts, 10, fact.content_fp)
    # Field 12 (sequence_id) is assigned by the subgraph mapping, not the client
    if fact.encrypted_embedding:
        _write_string(parts, 13, fact.encrypted_embedding)

    return b"".join(parts)


def encode_tombstone_protobuf(fact_id: str, owner: str) -> bytes:
    """Encode a tombstone payload for soft-deleting a fact on-chain.

    Sets ``decay_score=0`` and ``encrypted_blob`` to the bytes of
    ``"tombstone"`` (hex-encoded), matching the managed-service delete
    convention used by the TypeScript client.
    """
    tombstone = FactPayload(
        id=fact_id,
        timestamp=datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.000Z"),
        owner=owner,
        encrypted_blob=b"tombstone".hex(),
        blind_indices=[],
        decay_score=0.0,
        source="python_forget",
        content_fp="",
        agent_id="python-client",
    )
    return encode_fact_protobuf(tombstone)

File: rust/totalreclaw-core/src/protobuf.rs
//! Minimal protobuf encoder for TotalReclaw fact payloads.
//!
//! Hand-rolled wire format matching `mcp/src/subgraph/store.ts:encodeFactProtobuf()`.
//!
//! Field numbers match server/proto/totalreclaw.proto:
//!   1: id (string), 2: timestamp (string), 3: owner (string),
//!   4: encrypted_blob (bytes), 5: blind_indices (repeated string),
//!   6: decay_score (double), 7: is_active (bool), 8: version (int32),
//!   9: (removed in v3 — now encrypted inside field 4),
//!   10: content_fp (string),
//!   11: (removed in v3 — now encrypted inside field 4),
//!   12: sequence_id (int64, server-assigned), 13: encrypted_embedding (string)

/// A fact payload ready for protobuf encoding and on-chain submission.
#[derive(Debug, Clone)]
pub struct FactPayload {
    pub id: String,
    pub timestamp: String,
    pub owner: String,
    pub encrypted_blob_hex: String,
    pub blind_indices: Vec<String>,
    pub decay_score: f64,
    pub source: String,
    pub content_fp: String,
    pub agent_id: String,
    pub encrypted_embedding: Option<String>,
}

/// Encode a fact payload as minimal protobuf wire format.
pub fn encode_fact_protobuf(fact: &FactPayload) -> Vec<u8> {
    let mut buf = Vec::with_capacity(512);

    // Field 1: id (string)
    write_string(&mut buf, 1, &fact.id);
    // Field 2: timestamp (string)
    write_string(&mut buf, 2, &fact.timestamp);
    // Field 3: owner (string)
    write_string(&mut buf, 3, &fact.owner);
    // Field 4: encrypted_blob (bytes) — stored as hex, decode to raw bytes
    if let Ok(blob_bytes) = hex::decode(&fact.encrypted_blob_hex) {
        write_bytes(&mut buf, 4, &blob_bytes);
    }
    // Field 5: blind_indices (repeated string)
    for index in &fact.blind_indices {
        write_string(&mut buf, 5, index);
    }
    // Field 6: decay_score (double)
    write_double(&mut buf, 6, fact.decay_score);
    // Field 7: is_active (bool = varint 1)
    write_varint_field(&mut buf, 7, 1);
    // Field 8: version (int32 = varint 3)
    write_varint_field(&mut buf, 8, 3);
    // Fields 9 (source) and 11 (agent_id) removed in v3 — now encrypted inside field 4
    // Field 10: content_fp (string)
    write_string(&mut buf, 10, &fact.content_fp);
    // Field 12: sequence_id — assigned by subgraph, not set client-side
    // Field 13: encrypted_embedding (string)
    if let Some(ref emb) = fact.encrypted_embedding {
        write_string(&mut buf, 13, emb);
    }

    buf
}

/// Encode a tombstone protobuf for soft-deleting a fact.
pub fn encode_tombstone_protobuf(fact_id: &str, owner: &str) -> Vec<u8> {
    let mut buf = Vec::with_capacity(128);

    write_string(&mut buf, 1, fact_id);
    write_string(&mut buf, 2, &chrono::Utc::now().to_rfc3339());
    write_string(&mut buf, 3, owner);
    // Empty encrypted blob
    write_bytes(&mut buf, 4, &[]);
    // decay_score = 0 (tombstone signal)
    write_double(&mut buf, 6, 0.0);
    // is_active = false
    write_varint_field(&mut buf, 7, 0);
    write_varint_field(&mut buf, 8, 3);
    // Fields 9 (source) and 11 (agent_id) removed in v3

    buf
}

// ---------------------------------------------------------------------------
// Wire-format helpers
// ---------------------------------------------------------------------------

fn write_string(buf: &mut Vec<u8>, field: u32, value: &str) {
    if value.is_empty() {
        return;
    }
    let data = value.as_bytes();
    let key = (field << 3) | 2; // wire type 2 = length-delimited
    encode_varint(buf, key);
    encode_varint(buf, data.len() as u32);
    buf.extend_from_slice(data);
}

fn write_bytes(buf: &mut Vec<u8>, field: u32, value: &[u8]) {
    let key = (field << 3) | 2;
    encode_varint(buf, key);
    encode_varint(buf, value.len() as u32);
    buf.extend_from_slice(value);
}

fn write_double(buf: &mut Vec<u8>, field: u32, value: f64) {
    let key = (field << 3) | 1; // wire type 1 = 64-bit
    encode_varint(buf, key);
    buf.extend_from_slice(&value.to_le_bytes());
}

fn write_varint_field(buf: &mut Vec<u8>, field: u32, value: u32) {
    let key = (field << 3) | 0; // wire type 0 = varint
    encode_varint(buf, key);
    encode_varint(buf, value);
}

fn encode_varint(buf: &mut Vec<u8>, mut value: u32) {
    loop {
        if value <= 0x7f {
            buf.push(value as u8);
            break;
        }
        buf.push(((value & 0x7f) | 0x80) as u8);
        value >>= 7;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_varint_encoding() {
        let mut buf = Vec::new();
        encode_varint(&mut buf, 1);
        assert_eq!(buf, vec![1]);

        buf.clear();
        encode_varint(&mut buf, 300);
        assert_eq!(buf, vec![0xAC, 0x02]);
    }

    #[test]
    fn test_encode_fact_protobuf() {
        let payload = FactPayload {
            id: "test-id".into(),
            timestamp: "2026-01-01T00:00:00Z".into(),
            owner: "0xABCD".into(),
            encrypted_blob_hex: "deadbeef".into(),
            blind_indices: vec!["hash1".into(), "hash2".into()],
            decay_score: 0.8,
            source: "zeroclaw_fact".into(),
            content_fp: "fp123".into(),
            agent_id: "zeroclaw".into(),
            encrypted_embedding: None,
        };
        let encoded = encode_fact_protobuf(&payload);
        assert!(!encoded.is_empty());
        // Should contain the string "test-id" somewhere
        assert!(encoded.windows(7).any(|w| w == b"test-id"));
    }
}
